### PR TITLE
[`pyupgrade`] - add PEP646 Unpack conversion to `*` with fix (`UP044`)

### LINKED ISSUE #13891: pyupgrade rule for Unpack
It seems ruff is missing https://github.com/asottile/pyupgrade/pull/745 `G[Unpack[T]] => G[*T] ` to change UNPACK to asterisk if supported by the python version:

```
from typing import Unpack
def f(*x: Unpack[tuple[int, ...]]):
    return x

```
should be rewritten as
```
def f(*x: *tuple[int, ...]):
    return x
```

### PR DESCRIPTION
**Pull request by diceroll123**
## Summary

Adds `Unpack[...]` -> `*...` for Py311+

Closes #13891 

## Test Plan

`cargo test`

### CODE REVIEW

**Comment by charliermarsh** [comment]
Nice, thanks.

File: crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(NonPEP646Unpack, *range);
+
+    let inner = checker.locator().slice(slice.as_ref());
+
+    if checker.settings.preview.is_enabled() {
+        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+            format!("*{inner}"),
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs:78]: Unpacking anything other than a `Tuple` or `TypeVarTuple` is semantically incorrect but it isn't a syntax error. For example:

```py
def foo(*args: Unpack[int | str]) -> None:
```

Could we add a test asserting that the suggested fix doesn't suggest invalid syntax?

**Review by MichaReiser — CHANGES_REQUESTED** [review]
This is great. Thank you! 

Could you add one more test case for an unpacking where the target is a binary expression? This is is semantically invalid but it isn't a syntax error and we should at least avoid introducing a syntax error (I don't think we do, but we should make sure of it).

File: crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    if !checker.semantic().match_typing_expr(value, "Unpack") {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(NonPEP646Unpack, *range);
+
+    let inner = checker.locator().slice(slice.as_ref());
+
+    if checker.settings.preview.is_enabled() {
```
Comment (dhruvmanila) [inline crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs:76]: Why is this condition required when the rule itself is in preview?

Comment (diceroll123) [reply crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs:76]: Silly me!

File: crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    } = expr;
+
+    if !checker.semantic().match_typing_expr(value, "Unpack") {
+        return;
+    }
+
+    if slice.is_bin_op_expr() {
+        // fixing this would introduce a syntax error
+        return;
+    }
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs:75]: Applying the fix seems "safe enough":

```
def foo(*args: *int | str) -> None: pass  # not supported
```

Isn't invalid syntax

Comment (diceroll123) [reply crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs:75]: ah, true, it gives a TypeError actually.

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/pyupgrade/rules/use_pep646_unpack.rs:75]: I went ahead and converted it to an allowlist. I don't think any expressions other than name, subscript, and attribute are valid in this position.

**Review by MichaReiser — APPROVED** [review]

**Comment by sibarras** [comment]
It's giving SyntaxError when changing this:
```python3
from typing import TypedDict, Unpack


class KWs(TypedDict):
    """kws."""

    foo: str


def func(**kwargs: Unpack[KWs]) -> None:  # <-- original
    """Func."""
    _ = kwargs
```

to this:
```python3
from typing import TypedDict, Unpack


class KWs(TypedDict):
    """kws."""

    foo: str


def func(**kwargs: *KWs) -> None:  # <-- modified
    """Func."""
    _ = kwargs
```

**Comment by MichaReiser** [comment]
@sibarras see https://github.com/astral-sh/ruff/issues/14047

**Pull request merged**